File: frontend/src/queries/events.ts
/** The one connection that replaced the dashboard's timers.
 *
 *  Four things changed under this screen without anybody clicking, and each of
 *  them used to have its own `refetchInterval`: the property set every 4 s
 *  during a scan and 30 s otherwise, the availability batch and the geocoding
 *  batch every 800 ms each. A dashboard left open on a phone spent the day
 *  asking to be told nothing had happened, twice over if a second tab was open.
 *
 *  Now the backend says so once, down `GET /api/events`, and this module turns
 *  each frame into the cache write or the invalidation it means. The queries
 *  themselves are unchanged: a component still reads `useScanStatus()`, and it
 *  is the same key with the same shape in it — what moved is who decides when
 *  it is stale.
 *
 *  **`fetch`, not `EventSource`.** `EventSource` cannot send a header, and this
 *  route is behind the same optional `api_auth_token` gate as every other
 *  `/api` route (invariant 14). A stream that quietly stopped working the
 *  moment somebody set a token would be worse than no stream, and a second
 *  unauthenticated route to carry it would be worse still.
 *
 *  **The polling stays, as a fallback.** An old backend behind a new build, or
 *  anything between the browser and the process that will not carry a streaming
 *  response, has to leave a working dashboard rather than a frozen one. Three
 *  failed attempts to open the stream turn the intervals back on;
 *  `usePollingFallback()` is what the queries read to decide.
 */
import { useEffect, useSyncExternalStore } from "react";
import type { QueryClient } from "@tanstack/react-query";
import { useQueryClient } from "@tanstack/react-query";
import { authToken } from "../services/api";
import type { AvailabilityCheckProgress, GeocodeProgress, ScanStatus } from "../types";
import { keys } from "./keys";

/** First retry, doubling from here. Short enough that a backend restart is not
 *  something the user sits through. */
const RETRY_BASE_MS = 500;
/** …and capped well inside the ten seconds a reconnection is allowed to take. */
const RETRY_CAP_MS = 5000;
/** Failed *opens* before the timers come back. A stream that opened and then
 *  ended is not one of these — that is an ordinary reconnection, and treating
 *  it as a failure would put the polling back on every restart. */
const FAILURES_BEFORE_POLLING = 3;

/* ─────────────────────── is the stream carrying us? ─────────────────────── */

let pollingFallback = false;
const listeners = new Set<() => void>();

function setPollingFallback(on: boolean): void {
  if (on === pollingFallback) return;
  pollingFallback = on;
  for (const listener of listeners) listener();
}

function subscribeToFallback(listener: () => void): () => void {
  listeners.add(listener);
  return () => listeners.delete(listener);
}

/**
 * Should this query poll?
 *
 * `false` — the normal answer — means the stream is doing the job and an
 * interval would be pure duplication. It flips to `true` only when the stream
 * could not be opened at all, which is the one case where a timer is still the
 * difference between a live dashboard and a dead one.
 */
export function usePollingFallback(): boolean {
  return useSyncExternalStore(
    subscribeToFallback,
    () => pollingFallback,
    // Nothing has tried to connect yet wherever there is no browser to try
    // (a unit test rendering a component): poll, because that is the behaviour
    // that works without a stream.
    () => true,
  );
}

/* ────────────────────────────── the frames ────────────────────────────── */

export interface StreamEvent {
  topic: string;
  data: unknown;
}

/**
 * The complete frames in what has arrived so far, and the tail that is still
 * only half a frame.
 *
 * Split out and pure because this is the part that is easy to get subtly wrong
 * and impossible to see going wrong: a chunk boundary falls wherever TCP puts
 * it, so a frame arrives in two pieces often enough to matter and never in the
 * one run somebody watches by hand. Lines starting `:` are the server's
 * heartbeat — they carry nothing and exist so a dead socket is noticed.
 */
export function parseFrames(buffer: string): { events: StreamEvent[]; rest: string } {
  const parts = buffer.split("\n\n");
  const rest = parts.pop() ?? "";
  const events: StreamEvent[] = [];
  for (const block of parts) {
    let topic = "";
    const payload: string[] = [];
    for (const line of block.split("\n")) {
      if (line.startsWith("event:")) topic = line.slice(6).trim();
      else if (line.startsWith("data:")) payload.push(line.slice(5).trim());
    }
    if (!topic || payload.length === 0) continue;
    try {
      events.push({ topic, data: JSON.parse(payload.join("\n")) });
    } catch {
      // A frame this client cannot read is a frame it skips. The next one on
      // that topic is a whole snapshot too, so nothing stays wrong for long.
    }
  }
  return { events, rest };
}

/** What one frame means to the cache.
 *
 *  The three progress topics are written straight in rather than triggering a
 *  refetch: the frame *is* the answer that request would have come back with,
 *  serialised by the same response model, so asking again would be asking for
 *  what is already in hand. The health topic is the exception — it carries a
 *  fingerprint rather than the panel, because the panel is a thirty-day
 *  aggregate nobody should pay for every tick. */
export function applyEvent(client: QueryClient, { topic, data }: StreamEvent): void {
  switch (topic) {
    case "status":
      client.setQueryData<ScanStatus>(keys.scanStatus, data as ScanStatus);
      break;
    case "availability":
      client.setQueryData<AvailabilityCheckProgress>(
        keys.availabilityProgress,
        data as AvailabilityCheckProgress,
      );
      break;
    case "geocode":
      client.setQueryData<GeocodeProgress>(keys.geocodeProgress, data as GeocodeProgress);
      break;
    case "health":
      // A blocked scan writes no property, so the fingerprint the grid watches
      // does not move and this is the only thing that says the portal stopped
      // answering. The profile rows carry the same state as the panel.
      void client.invalidateQueries({ queryKey: keys.scraperHealth });
      void client.invalidateQueries({ queryKey: keys.profiles });
      break;
  }
}

/* ──────────────────────────── the connection ──────────────────────────── */

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

/**
 * Hold one stream open until it ends, applying what comes down it.
 *
 * Returns whether it ever *opened*. That distinction is the whole of the
 * fallback rule: a stream that opened and later ended is a reconnection, and a
 * stream that never opened three times running is a backend that cannot serve
 * one.
 */
async function readStream(client: QueryClient, signal: AbortSignal): Promise<boolean> {
  const token = authToken.get();
  const response = await fetch("/api/events", {
    signal,
    headers: {
      Accept: "text/event-stream",
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
  });
  // A 401 is left to the ordinary request path: the fallback poll below will
  // make one, and the auth gate it raises is the same prompt as always. There
  // is no second login here to keep in step with the first.
  if (!response.ok || !response.body) return false;
  setPollingFallback(false);

  const reader = response.body.pipeThrough(new TextDecoderStream()).getReader();
  let buffer = "";
  try {
    for (;;) {
      const { value, done } = await reader.read();
      if (done) break;
      buffer += value;
      const { events, rest } = parseFrames(buffer);
      buffer = rest;
      for (const event of events) applyEvent(client, event);
    }
  } finally {
    void reader.cancel().catch(() => {});
  }
  return true;
}

/**
 * Subscribe for as long as the dashboard is mounted.
 *
 * Mounted once, on the layout route, so it survives the URL moving between the
 * grid, a property and Settings — a stream reopened on every navigation would
 * be four timers replaced by a reconnection storm.
 */
export function useEventStream(): void {
  const client = useQueryClient();

  useEffect(() => {
    const controller = new AbortController();
    let stopped = false;
    let failures = 0;

    void (async () => {
      while (!stopped) {
        try {
          const opened = await readStream(client, controller.signal);
          failures = opened ? 0 : failures + 1;
        } catch {
          // Aborted on unmount, or the socket dropped. Both end the same way:
          // count it, back off, try again.
          failures += 1;
        }
        if (stopped) return;
        if (failures >= FAILURES_BEFORE_POLLING) setPollingFallback(true);
        await sleep(Math.min(RETRY_BASE_MS * 2 ** (failures - 1), RETRY_CAP_MS));
      }
    })();

    return () => {
      stopped = true;
      controller.abort();
    };
  }, [client]);
}

File: backend/app/services/events.py
"""One outbound stream carrying everything the dashboard used to ask for on a timer.

Four things change under the dashboard without anyone clicking: the property set
(a scan lands, a second tab hides a row), the scan's own progress, the
availability batch's, and the geocoding batch's. Each of them had its own timer
in the browser — 800 ms for the two batches, 4 s during a scan and 30 s
otherwise for the grid — which meant a dashboard sitting open on a phone all day
made thousands of requests to be told nothing had happened, and a browser that
had two tabs open made twice as many.

What replaces them is this module and `routers/events.py`: **one sampler for the
whole process**, publishing to every connected browser at once. The shape is
deliberately a sampler rather than a set of hooks in the producers:

* the producers are threads. `scan_state`, `_scan_progress`, the two batch
  progress dicts — all of them are module-level dicts written from a worker
  thread and copied out on read, which is the existing design and a good one.
  Reaching into an event loop from those threads to publish would put an
  `asyncio` handoff into the middle of a scan; reading their snapshots from here
  costs them nothing and cannot break them.
* the fingerprint has no producer at all. `properties_version` is a property of
  the database, moved by a scan, by the user hiding a row, and by a second tab
  doing either. Nothing in the code is in a position to announce it; the only
  honest way to know is to look.

So the total cost of knowing is **one tick per second, for the whole machine,
and only while at least one browser is connected** — against one HTTP request
per timer per tab before. The sampler stops when the last subscriber leaves.

Nothing here holds a database session open: each tick opens one, reads two small
aggregates, and closes it. A session held for the life of an SSE connection
would pin a SQLite connection from the pool for as long as a tab stayed open,
which on the four-connection default is four idle tabs and no scan.

Invariant 14 is untouched. This is a `GET` on the existing origin that the
browser opens and the server writes to: no inbound port, no webhook, no
callback address. It is behind the same `require_api_token` middleware as every
other `/api` route, which is why the client reads it with `fetch` and an
`Authorization` header rather than with `EventSource` — `EventSource` cannot
send one.
"""

from __future__ import annotations

import asyncio
import json
import logging
from collections.abc import AsyncGenerator
from typing import Any

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from .. import schemas
from ..database import SessionLocal
from ..models import PriceHistory, Property, SearchProfile

logger = logging.getLogger(__name__)

# How often the sampler looks. One second is the resolution of the fastest thing
# it reports (the batch progress bars, which used to poll at 800 ms) and is well
# under the cadence of the slowest (a scan page, seconds apart at best).
TICK_SECONDS = 1.0

# A line down an idle connection, so a client that has silently lost the server
# — standby, a dropped Wi-Fi, a proxy that reaps quiet sockets — finds out
# rather than sitting on a stream that will never speak again.
HEARTBEAT_SECONDS = 10.0

# Frames a client may fall behind by before it is resynchronised wholesale.
# Every frame is a complete snapshot of its topic, so the recovery is simply to
# send the current one rather than to replay what was missed.
_QUEUE_LIMIT = 32


def frame(topic: str, payload: Any) -> str:
    """One Server-Sent Events frame: a named event and its JSON body."""
    return f"event: {topic}\ndata: {json.dumps(payload, default=str)}\n\n"


def properties_version(db: Session) -> str:
    """A cheap fingerprint of the property set: "did anything change?".

    The dashboard used to answer that question by re-downloading every filtered
    property — market position, match score, deal score and provenance computed
    for each — every 30 seconds, and every 4 seconds during a scan. Now the
    sampler above reads this string once for the whole machine and the grid is
    re-read only when it moves.

    Two small aggregates. The per-status counts are what make it trustworthy: a
    new property raises one, and every lifecycle transition — a scan marking
    rows `gone`, the user hiding one or marking it `sold` — moves a row from one
    bucket to another, which the grid must reflect because those rows enter or
    leave it. Grouping rather than a single total is the difference: hiding a
    property leaves `count(*)` exactly where it was. `max(last_seen_at)` catches
    a scan that re-found existing listings and changed nothing else, and the
    newest price-history id catches a price change on a property already shown.

    Not a perfect change detector, and it does not try to be: editing notes or
    toggling a favourite moves none of these. Those are user actions, and the
    client that performs one already refreshes on the response.
    """
    rows = db.execute(
        select(
            Property.status,
            func.count(Property.id),
            func.max(Property.id),
            func.max(Property.last_seen_at),
        ).group_by(Property.status)
    ).all()
    last_price = db.execute(select(func.max(PriceHistory.id))).scalar()
    buckets = ";".join(
        f"{status}:{count}:{max_id}:{last_seen}"
        for status, count, max_id, last_seen in sorted(rows, key=lambda r: r[0] or "")
    )
    return f"{buckets}|{last_price}"


def health_version(db: Session) -> str:
    """A fingerprint of the scraping health, for the same reason as above.

    A blocked scan is the one that changes nothing else: it writes no property,
    so `properties_version` sits exactly where it was and the health panel would
    have kept showing "ok" until something unrelated moved. The outcome and the
    failure streak are the two columns the panel and the profile rows read, so
    those are what is watched.
    """
    rows = db.execute(
        select(
            SearchProfile.id,
            SearchProfile.last_run_status,
            SearchProfile.consecutive_failures,
            SearchProfile.last_run_at,
        ).order_by(SearchProfile.id)
    ).all()
    return ";".join(f"{pid}:{status}:{streak}:{ran}" for pid, status, streak, ran in rows)


def collect_snapshot() -> dict[str, Any]:
    """Everything the stream reports, as it stands right now.

    Serialised through the same response models as the REST routes that used to
    be polled, so a client receives byte-identical shapes whichever way it
    learns. A payload hand-assembled here would drift from the schema the
    generated types are built from, and the drift would surface as an
    `undefined` in the browser rather than as a failing gate.

    Synchronous on purpose: the caller runs it in a worker thread, because the
    two queries below are SQLite reads and blocking the event loop on them would
    stall every other connection the process is serving.
    """
    from ..config import load_settings
    from . import scheduler
    from .availability_check import get_prop_check_progress
    from .geocoder import get_geocode_progress
    from .scanner import get_scan_progress, scan_state

    with SessionLocal() as db:
        data_version = properties_version(db)
        health = health_version(db)

    status = schemas.ScraperStatusOut(
        **scan_state,
        next_auto_run=scheduler.next_run_time(),
        paused=bool(load_settings().get("scanning_paused")),
        data_version=data_version,
        progress=schemas.ScanProgressOut(**get_scan_progress()),
    )
    return {
        "status": status.model_dump(mode="json"),
        "availability": schemas.AvailabilityCheckProgressOut(
            **get_prop_check_progress()
        ).model_dump(mode="json"),
        "geocode": schemas.GeocodeProgressOut(**get_geocode_progress()).model_dump(mode="json"),
        "health": {"version": health},
    }


class EventHub:
    """The subscribers, the last snapshot of each topic, and the sampler.

    One instance per process (`hub`, below). It publishes a topic only when its
    payload differs from the last one sent, which is what makes a second of
    resolution affordable: an idle dashboard receives a heartbeat every ten
    seconds and nothing else.
    """

    def __init__(self) -> None:
        self._queues: set[asyncio.Queue[str]] = set()
        self._latest: dict[str, Any] = {}
        self._sampler: asyncio.Task[None] | None = None

    # -- publishing -------------------------------------------------------

    def publish(self, topic: str, payload: Any) -> None:
        """Send a topic to every subscriber, if it has actually moved."""
        if self._latest.get(topic) == payload:
            return
        self._latest[topic] = payload
        text = frame(topic, payload)
        for queue in list(self._queues):
            if queue.full():
                # This client is not reading fast enough. Replaying the backlog
                # would only make it further behind, and every frame is a whole
                # snapshot anyway, so drop what is queued and hand it the world
                # as it stands.
                self._resync(queue)
                continue
            queue.put_nowait(text)

    def _resync(self, queue: asyncio.Queue[str]) -> None:
        while not queue.empty():
            queue.get_nowait()
        for topic, payload in self._latest.items():
            if queue.full():
                return
            queue.put_nowait(frame(topic, payload))

    async def _sample(self) -> None:
        snapshot = await asyncio.to_thread(collect_snapshot)
        for topic, payload in snapshot.items():
            self.publish(topic, payload)

    async def _run(self) -> None:
        """Sample for as long as somebody is listening, then stop."""
        try:
            while self._queues:
                try:
                    await self._sample()
                except asyncio.CancelledError:
                    raise
                except Exception:
                    # A sampler that dies takes every connected dashboard's
                    # updates with it, silently. One bad tick is not worth that.
                    logger.exception("event sampler tick failed")
                await asyncio.sleep(TICK_SECONDS)
        finally:
            self._sampler = None

    def _ensure_sampler(self) -> None:
        if self._sampler is None or self._sampler.done():
            self._sampler = asyncio.create_task(self._run())

    # -- subscribing ------------------------------------------------------

    async def subscribe(self) -> AsyncGenerator[str, None]:
        """One connection's worth of frames, until the client goes away.

        The current state of every topic is sent first, which is what makes a
        reconnection a resynchronisation: a browser that lost the stream while
        the backend restarted is fully up to date on the first frame of the new
        one, without asking for anything.
        """
        queue: asyncio.Queue[str] = asyncio.Queue(maxsize=_QUEUE_LIMIT)
        self._queues.add(queue)
        try:
            if not self._latest:
                # First subscriber of this process: sample now rather than make
                # it wait a tick to learn anything at all.
                await self._sample()
            self._ensure_sampler()
            # Anything that sample just queued is already in the snapshot below,
            # and sending both would open every connection with each topic
            # twice — which for a dashboard is a duplicate invalidation and for
            # a test is an idle stream that looks anything but.
            while not queue.empty():
                queue.get_nowait()
            for topic, payload in list(self._latest.items()):
                yield frame(topic, payload)
            while True:
                try:
                    yield await asyncio.wait_for(queue.get(), HEARTBEAT_SECONDS)
                except TimeoutError:
                    yield ": ping\n\n"
        finally:
            self._queues.discard(queue)

    # -- shutdown ---------------------------------------------------------

    async def aclose(self) -> None:
        """Stop the sampler. Called when the application shuts down."""
        task = self._sampler
        self._queues.clear()
        self._sampler = None
        if task is not None and not task.done():
            task.cancel()
            try:
                await task
            except (asyncio.CancelledError, Exception):
                pass


hub = EventHub()
